File: code/HuaWeiNote/app/src/test/java/com/fan/hwnote/app/model/TodoRepositoryTest.kt
package com.fan.hwnote.app.model

import androidx.test.core.app.ApplicationProvider
import com.fan.hwnote.app.model.entity.RepeatType
import com.fan.hwnote.app.model.entity.Todo
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class TodoRepositoryTest {

    @Before
    fun setUp() {
        val ctx = ApplicationProvider.getApplicationContext<android.content.Context>()
        ctx.deleteDatabase("hwnote.db")
        TodoRepository.init(ctx)
    }

    @Test
    fun `insert returns positive id and getById retrieves it`() = runBlocking {
        val todo = Todo.new().copy(title = "买牛奶")
        val id = TodoRepository.insert(todo)
        assertTrue(id > 0)
        val fetched = TodoRepository.getById(id)
        assertNotNull(fetched)
        assertEquals("买牛奶", fetched!!.title)
        assertEquals(false, fetched.isCompleted)
    }

    @Test
    fun `update modifies fields`() = runBlocking {
        val id = TodoRepository.insert(Todo.new().copy(title = "旧标题"))
        val old = TodoRepository.getById(id)!!
        TodoRepository.update(old.copy(title = "新标题", isImportant = true))
        val updated = TodoRepository.getById(id)!!
        assertEquals("新标题", updated.title)
        assertEquals(true, updated.isImportant)
    }

    @Test
    fun `list returns non-deleted todos`() = runBlocking {
        TodoRepository.insert(Todo.new().copy(title = "正常"))
        TodoRepository.insert(Todo.new().copy(title = "已删", deletedAt = 1L))
        val list = TodoRepository.list()
        assertEquals(1, list.size)
        assertEquals("正常", list[0].title)
    }

    @Test
    fun `count returns correct number`() = runBlocking {
        TodoRepository.insert(Todo.new().copy(title = "A"))
        TodoRepository.insert(Todo.new().copy(title = "B"))
        assertEquals(2, TodoRepository.count())
    }

    @Test
    fun `softDelete sets deletedAt`() = runBlocking {
        val id = TodoRepository.insert(Todo.new().copy(title = "待删"))
        TodoRepository.softDelete(id)
        val todo = TodoRepository.getById(id)!!
        assertTrue(todo.deletedAt > 0)
    }

    @Test
    fun `restore clears deletedAt`() = runBlocking {
        val id = TodoRepository.insert(Todo.new().copy(title = "待恢复"))
        TodoRepository.softDelete(id)
        TodoRepository.restore(id)
        val todo = TodoRepository.getById(id)!!
        assertEquals(0L, todo.deletedAt)
    }

    @Test
    fun `deletePermanently removes from database`() = runBlocking {
        val id = TodoRepository.insert(Todo.new().copy(title = "彻底删"))
        TodoRepository.deletePermanently(id)
        assertNull(TodoRepository.getById(id))
    }

    @Test
    fun `purgeExpired deletes old soft-deleted todos`() = runBlocking {
        val now = System.currentTimeMillis()
        val oldTime = now - 31L * 24 * 60 * 60 * 1000
        TodoRepository.insert(Todo.new().copy(title = "旧删除", deletedAt = oldTime))
        TodoRepository.insert(Todo.new().copy(title = "新删除", deletedAt = now))
        val purged = TodoRepository.purgeExpired(now)
        assertEquals(1, purged)
    }
}

File: code/HuaweiNoteFlutter/test/repositories/todo_repository_test.dart
// test/repositories/todo_repository_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:sqflite_common_ffi/sqflite_ffi.dart';
import 'package:hwnote/db/database_helper.dart';
import 'package:hwnote/repositories/todo_repository.dart';
import 'package:hwnote/models/todo.dart';

void main() {
  sqfliteFfiInit();
  databaseFactory = databaseFactoryFfi;

  late DatabaseHelper dbHelper;
  late TodoRepository repo;

  setUp(() async {
    dbHelper = DatabaseHelper(inMemory: true);
    await dbHelper.database;
    repo = TodoRepository(dbHelper);
  });

  tearDown(() async => await dbHelper.close());

  group('TodoRepository', () {
    test('insert and getById', () async {
      final todo = Todo.newTodo().copyWith(title: 'Buy milk');
      final id = await repo.insert(todo);
      expect(id, greaterThan(0));
      final loaded = await repo.getById(id);
      expect(loaded, isNotNull);
      expect(loaded!.title, 'Buy milk');
    });

    test('update changes fields', () async {
      final id = await repo.insert(Todo.newTodo().copyWith(title: 'Old'));
      final loaded = (await repo.getById(id))!;
      await repo.update(loaded.copyWith(title: 'New'));
      final updated = await repo.getById(id);
      expect(updated!.title, 'New');
    });

    test('list returns non-deleted todos', () async {
      await repo.insert(Todo.newTodo().copyWith(title: 'A'));
      await repo.insert(Todo.newTodo().copyWith(title: 'B'));
      final list = await repo.list();
      expect(list.length, 2);
    });

    test('softDelete marks as deleted', () async {
      final id = await repo.insert(Todo.newTodo().copyWith(title: 'Del'));
      await repo.softDelete(id);
      final list = await repo.list();
      expect(list, isEmpty);
    });

    test('restore brings back deleted todo', () async {
      final id = await repo.insert(Todo.newTodo().copyWith(title: 'Restore'));
      await repo.softDelete(id);
      await repo.restore(id);
      final list = await repo.list();
      expect(list.length, 1);
    });

    test('completeTodo sets is_completed for non-repeating', () async {
      final id = await repo.insert(Todo.newTodo().copyWith(title: 'Done'));
      await repo.completeTodo(id);
      final loaded = await repo.getById(id);
      expect(loaded!.isCompleted, true);
    });

    test('completeTodo advances remindAt for repeating', () async {
      final now = DateTime.now().millisecondsSinceEpoch;
      final id = await repo.insert(Todo.newTodo().copyWith(
        title: 'Repeat',
        remindAt: now - 1000,
        repeatType: RepeatType.daily,
      ));
      await repo.completeTodo(id);
      final loaded = await repo.getById(id);
      expect(loaded!.isCompleted, false);
      expect(loaded.remindAt, greaterThan(now));
    });

    test('count returns correct number', () async {
      await repo.insert(Todo.newTodo().copyWith(title: 'A'));
      await repo.insert(Todo.newTodo().copyWith(title: 'B'));
      expect(await repo.count(), 2);
    });

    test('advanceRemindAt advances past now', () {
      final past = DateTime.now().millisecondsSinceEpoch - 86400000 * 3;
      final result = TodoRepository.advanceRemindAt(past, RepeatType.daily);
      expect(result, greaterThan(DateTime.now().millisecondsSinceEpoch));
    });

    test('purgeExpired removes old soft-deleted todos', () async {
      final id = await repo.insert(Todo.newTodo().copyWith(title: 'Old'));
      await repo.softDelete(id);
      final db = await dbHelper.database;
      final longAgo =
          DateTime.now().millisecondsSinceEpoch - 31 * 24 * 60 * 60 * 1000;
      await db.update('todos', {'deleted_at': longAgo},
          where: 'id = ?', whereArgs: [id]);
      final purged = await repo.purgeExpired();
      expect(purged, 1);
      expect(await repo.getById(id), isNull);
    });
  });
}
